### PR TITLE
WIP: Add `not_exhaustive_enough´ lint

### LINKED ISSUE #5557: New lint: warn about missing variants/fields for non-exhaustive enums/structs
As suggested in the [non_exhaustive RFC](https://github.com/rust-lang/rfcs/blob/master/text/2008-non-exhaustive.md#unresolved-questions), when using non-exhaustive enums and structs in patterns, the lint would warn the user for missing variants or fields despite having a wildcard arm or a rest pattern.

This can be useful when new fields/variants are added by the upstream crate so that they don't go unnoticed. 

### Enums
Given the following definition
```rust
#[non_exhaustive]
pub enum E {
    First,
    Second,
    Third,
}
```
The following use should be linted:
```rust
    match e {
        E::First => {}
        E::Second => {}
        _ => {}
    }
```

### Structs
Given the following definition
```rust
#[derive(Default)]
#[non_exhaustive]
pub struct S {
    pub a: i32,
    pub b: i32,
    pub c: i32,
}
```
The following cases should be linted
```rust
    let S { a: _, b: _, .. } = S::default();
    match S::default() {
        S { a: 42, b: 21, .. } => {}
        S { a: _, b: _, .. } => {}
    }
    if let S { a: 42, b: _, .. } = S::default() {}
    while let S { a: 42, b: _, .. } = S::default() {
        break;
    }
    let v = vec![S::default()];
    for S { a: _, b: _, .. } in v {}
```
In addition to let destructuring in function parameters
```rust
pub fn take_s(S { a, b, .. }: S) -> (i32, i32) {
    (a, b)
}
```

A struct variant of an enum should be linted in the same cases if it's the only variant of the enum. If it's not the only member, it can be linted only in the *match*, *if let* and *while let* cases.

### Tuple structs
Tuple structs can't be used in patterns if there are private fields using the `T()` syntax, but the `T{0: _, 1: _}` syntax can be used an so it should be linted like with structs.

Given the following definition
```rust
#[derive(Default)]
#[non_exhaustive]
pub struct T(pub i32, pub i32, pub i32);
```
The following cases should be linted
```rust
    let T { 0: _, 1: _, .. } = T::default();
    match T::default() {
        T { 0: 42, 1: 21, .. } => {}
        T { 0: _, 1: _, .. } => {}
    }
    if let T { 0: 42, 1: _, .. } = T::default() {}
    while let T { 0: 42, 1: _, .. } = T::default() {
        break;
    }
    let v = vec![T::default()];
    for T { 0: _, 1: _, .. } in v {}
```
In addition to let destructuring in function parameters
```rust
pub fn take_t(T { 0: a, 1: b, .. }: T) -> (i32, i32) {
    (a, b)
}
```

### Notes
* The non-exhaustive data structure must come from another crate (otherwise the restrictions associated to non-exhaustive data structures do not apply).
* The lint should not trigger if only the rest pattern is used, so at least one field should be used in the pattern.
* The lint should be allow-by-default.
* Although described in the RFC, FRU is not supported currently so constructing a non-exhaustive struct is not possible.
* Unit structs are not candidates for this lint because new fields can't be added.

### PR DESCRIPTION
**Pull request by jrqc**
r? @ebroto 
Fixes #5557
changelog: new lint `not_exhaustive_enough`

### CODE REVIEW

File: clippy_lints/src/not_exhaustive_enough.rs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         _ => {}
+    ///     }
+    /// // Good
+    /// match e {
+    ///         E::First => {}
+    ///         E::Second => {}
+    ///         E::Third => {}
+    ///     }
```
Comment (Rantanen) [inline clippy_lints/src/not_exhaustive_enough.rs:35]: The point of `#[non_exhaustive]` is that the wildcard match is always required:
```suggestion
    ///         E::Third => {}
    ///         _ => {}
    ///     }
```

File: clippy_lints/src/not_exhaustive_enough.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+}
+
+fn check_path_pat<'tcx>(cx: &LateContext<'tcx>, arms: &[Arm<'_>], e: &'tcx Expr<'_>) {
+    let missing_variants = get_missing_variants(cx, arms, e);
+    span_lint_and_sugg(
+        cx,
+        NOT_EXHAUSTIVE_ENOUGH,
+        e.span,
+        "Enum not exhaustive enough",
```
Comment (ebroto) [inline clippy_lints/src/not_exhaustive_enough.rs:84]:
```suggestion
        "enum match is not exhaustive enough",
```
We generally avoid capitalizing the first letter

**Review by ebroto — CHANGES_REQUESTED** [review]
Thanks for the effort you've put into this! 

I did a first round of review, don't hesitate to ask for clarification if needed.

**Comment by ebroto** [comment]
Also, my apologies for the delay in reviewing, I did not have much free time lately.

**Comment by Rantanen** [comment]
I'm slightly concerned that the full use of this lint will be held back by inability to distinguish between:
- I want to only match these two items and don't care about the rest.
- I want to match everything, but need to use wildcard for `non_exhaustive`.

```rust
match io_error.kind() {
  ErrorKind::TimedOut => self.retry(),
  _ => return Err(io_error),  // This is intended to match all known and unknown kinds.
}
```

```rust
if let Some(Foo { x, .. /* don't care for the rest */ }) = maybe_foo {
  ..
}
```

These issues can always be ignored case-by-case with `#[allow(..)]` attributes on the statements, but if that has to be done frequently the added noise of said attributes might outweigh the benefits.

I don't really have any good solutions to this problem and I do feel having the lint as it is implemented here is better than not having it at all. :)

**Comment by jrqc** [comment]
Thank you for the review! I will make the necessary changes as soon as possible

**Comment by ebroto** [comment]
@Rantanen that's a very good point. 

The idea behind the lint is to warn in case new fields/variants are added, so it should target those cases where the user wanted to be as exhaustive as possible. But we don't know how the data structure looked like when the user wrote the code, so we can only approximate.

Some possible heuristics could be (although I'm not happy with any of them TBH):
* Warn only if a majority of the fields/variants are already included (i.e. >= n/2+1).
* Do not warn if only one field/variant is included (without taking wildcard/rest pattern into account).

**Comment by ebroto** [comment]
If the lint turns out to be more annoying than helpful we can always turn it into a `restriction` lint.

**Comment by giraffate** [comment]
ping from triage @jrqc. There seem to be some small fixes left to be done. Could you have any update on this?

**Comment by jrqc** [comment]
Sorry about the delayed response @ebroto @giraffate 
I will update it today

**Comment by jrqc** [comment]
@ebroto Made the necessary changes! Hope I didn't miss anything 
Again sorry about the late edit

**Comment by ebroto** [comment]
>Sorry about the delayed response

No problem, thanks for sticking with it! I will try to review in the coming days.

**Comment by Manishearth** [comment]
I'll review this when I get a chance, but I'm wondering: we're replicating a lot of the exhaustiveness checking logic in rustc, yeah? Can we use it, or make this lint a lint in rustc?

**Comment by Manishearth** [comment]
@flip1995 what are your opinions on the above^ ?

**Comment by flip1995** [comment]
I don't really have an opinion on this, since I didn't review any of the exhaustiveness PRs AFAICR. I just can give a general answer: If we can share code with rustc, we should do that. But I don't know how code in Clippy or rustc looks like for exhaustiveness checks.

**Comment by Manishearth** [comment]
@flip1995 so the thing is that rustc already has exhaustiveness checks, and this is a lint that was proposed in a Rust RFC, so it would both be _easy_ to implement in rustc and _make sense_. It feels weird to reimplement exhaustiveness checking in a partial way outside of rustc when we can have a perfect version of this lint in rustc.

**Comment by flip1995** [comment]
Yeah, agreed.

**Comment by giraffate** [comment]
ping from triage @Manishearth. Do we still need discussion here?

**Comment by Manishearth** [comment]
Yeah, sorry, I think we've discussed this and we shouldn't be doing this in clippy. It's worth filing a bug on rust-lang/rust to add such an allow-default lint. Sorry @jrqc!

**Comment by jplatte** [comment]
I can't find an issue about this lint at rust-lang/rust, did anybody end up creating one? Also, given that this won't be implemented in clippy, somebody should close the associated issue (#5557).

**Pull request closed without merging**